pandas/io/parsers.py

```diff
@@ -3098,15 +3110,21 @@ def _clean_na_values(na_values, keep_default_na=True):
         na_fvalues = set()
     elif isinstance(na_values, dict):
         na_values = na_values.copy()  # Prevent aliasing.
-        if keep_default_na:
-            for k, v in compat.iteritems(na_values):
-                if not is_list_like(v):
-                    v = [v]
+
+        # Convert the values in the na_values dictionary
+        # into array-likes for further use. This is also
+        # where we append the default NaN values, provided
+        # that `keep_default_na=True`.
+        for k, v in compat.iteritems(na_values):
+            if not is_list_like(v):
+                v = [v]
+
+            if keep_default_na:
                 v = set(v) | _NA_VALUES
-                na_values[k] = v
-        na_fvalues = dict(
-            (k, _floatify_na_values(v)) for k, v in na_values.items()  # noqa
-        )
+
+            na_values[k] = v
```

Comment:
this is an anti-pattern to modify the dict that you are iterating. can you create you create a new one here?

Reply:
@jreback : That's why up above, they do `na_values = na_values.copy()`.

Reply:
yeah you should create a new empty dict and then assign to it (it doesn't have to be a copy). iterating and modifying is a no-no.

Reply:
I've decided to create a copy because at the bottom, we return `na_values` and `na_fvalues` regardless of the logic branch.  However, I'll iterate over the `old_na_values` instead.